### PR TITLE
Location Field: Add Typescript, i18n, and migrate to React Hooks

### PR DESCRIPTION
**Pull request by miles-grant-ibigroup**
Only works in OTP-RR with https://github.com/opentripplanner/otp-react-redux/pull/540

Blocked by #326 which adds typescript to the geocoder

Migrates location-field to typescript and moves it to a functional component with React hooks in the process.

Looking for feedback on:
- Some types might be a bit messy
- Some further optimizations might be possible now that we're moved to hooks

### CODE REVIEW

**Comment by demory** [comment]
I finally have this working within a local OTP-RR setup (applying OTP-RR PR 540 and using the copy-into-node-modules approach) -- everything looks good except when I search for an unrecognized string it says "No results found for {input}" without expanding the input. Is this an issue w/ the code or perhaps something w/ my setup?
image

**Review by demory — APPROVED** [review]

File: packages/location-field/src/index.tsx [hunk 1]
```diff
@@ -0,0 +1,758 @@
...
+function DefaultLocationIcon({
+  locationType
+}: {
+  locationType: string;
+}): React.ReactElement {
+  return <LocationIcon size={13} type={locationType} />;
+}
+
+const LocationField = ({
+  autoFocus = false,
```
Comment (binh-dam-ibigroup) [inline packages/location-field/src/index.tsx:41]: Sort props

File: packages/location-field/src/index.tsx [hunk 2]
```diff
@@ -0,0 +1,758 @@
...
+  GeocodedOptionIcon,
+  Option,
+  TransitStopOption,
+  UserLocationIcon
+} from "./options";
+import * as S from "./styled";
+import { generateLabel, getCombinedLabel } from "./utils";
+
+// FIXME have a better key generator for options
+let optionKey = 0;
```
Comment (binh-dam-ibigroup) [inline packages/location-field/src/index.tsx:30]: Can this var should be converted into a state?

File: packages/location-field/src/index.tsx [hunk 3]
```diff
@@ -0,0 +1,758 @@
...
+          // If no features found in response, default to empty array.
+          let geocodedFeatures = result && result.features;
+          if (!geocodedFeatures) {
+            // Get the Pelias error message if exists.
+            // TODO: determine how other geocoders return error messages.
+            const errorMessage =
+              result &&
+              result.results &&
+              result.results.error &&
+              result.results.error.message;
```
Comment (binh-dam-ibigroup) [inline packages/location-field/src/index.tsx:134]: Use `?.` syntax.

File: packages/location-field/src/index.tsx [hunk 4]
```diff
@@ -0,0 +1,758 @@
...
+      .autocomplete({ text })
+      // TODO: Better type?
+      .then(
+        (result: {
+          features: any;
+          results: { error: { message: string } };
+        }) => {
+          let message: string;
+          // If no features found in response, default to empty array.
+          let geocodedFeatures = result && result.features;
```
Comment (binh-dam-ibigroup) [inline packages/location-field/src/index.tsx:126]: Use `?.` syntax.

File: packages/location-field/src/index.tsx [hunk 5]
```diff
@@ -0,0 +1,758 @@
...
+            const errorMessage =
+              result &&
+              result.results &&
+              result.results.error &&
+              result.results.error.message;
+            // If the result did not contain a list of features, add special note.
+            message = intl.formatMessage({
+              id: "otpUi.LocationField.geocoderUnreachable"
+            });
+            if (errorMessage) message += ` (${errorMessage})`;
```
Comment (binh-dam-ibigroup) [inline packages/location-field/src/index.tsx:139]: Remove this string concat. Instead, include the stuff in parenthesis as a conditional message inside the localized string, pass `errorMessage` as a parameter when calling `intl.formatMessage()`.

File: packages/location-field/src/index.tsx [hunk 6]
```diff
@@ -0,0 +1,758 @@
...
+        console.error(err);
+      });
+  });
+
+  const getFormControlClassname = () => {
+    return `${locationType}-form-control`;
+  };
+
+  const setLocation = (newLocation: Location, resultType: ResultType) => {
+    onLocationSelected(intl, { locationType, location: newLocation, resultType });
```
Comment (binh-dam-ibigroup) [inline packages/location-field/src/index.tsx:161]: Sort attributes

File: packages/location-field/src/index.tsx [hunk 7]
```diff
@@ -0,0 +1,758 @@
...
+
+  const useCurrentLocation = () => {
+    const newLocation = coreUtils.map.currentPositionToLocation(
+      currentPosition
+    );
+    if (newLocation) {
+      // If geolocation is successful (i.e., user has granted app geolocation
+      // permission and coords exist), set location.
+      onLocationSelected(intl, {
+        locationType,
```
Comment (binh-dam-ibigroup) [inline packages/location-field/src/index.tsx:173]: Sort attributes.

File: packages/location-field/src/index.tsx [hunk 8]
```diff
@@ -0,0 +1,758 @@
...
+
+  const geocodeSearch = text => {
+    if (!text) {
+      console.warn("No text entry provided for geocode search.");
+      return;
+    }
+    getGeocoder(geocoderConfig)
+      .search({ text })
+      .then(result => {
+        if (result.features && result.features.length > 0) {
```
Comment (binh-dam-ibigroup) [inline packages/location-field/src/index.tsx:307]: Use `?.` notation.

File: packages/location-field/src/index.tsx [hunk 9]
```diff
@@ -0,0 +1,758 @@
...
+          // populate the sessionSearches array.
+          addLocationSearch({ location: geocodedLocation });
+        });
+    };
+
+    // Add to the selection handler lookup (for use in onKeyDown)
+    locationSelectedLookup[itemIndex] = locationSelected;
+
+    // Extract GTFS/POI info and assign to class
+    const { layer, source, id } = feature.properties;
```
Comment (binh-dam-ibigroup) [inline packages/location-field/src/index.tsx:347]: Sort props.

File: packages/location-field/src/index.tsx [hunk 10]
```diff
@@ -0,0 +1,758 @@
...
+      ),
+      otherFeatures.map(feature => renderFeature(itemIndex++, feature))
+    );
+  }
+
+  /* 2) Process nearby transit stop options */
+  if (nearbyStops.length > 0 && !suppressNearby) {
+    // Add the menu sub-heading (not a selectable item)
+    menuItems.push(
+      <S.MenuItem header centeredText key="ns-header">
```
Comment (binh-dam-ibigroup) [inline packages/location-field/src/index.tsx:485]: Sort props (multiple instances).

File: packages/location-field/src/index.tsx [hunk 11]
```diff
@@ -0,0 +1,758 @@
...
+        const locationSelected = () => {
+          setLocation(sessionLocation, "SESSION");
+        };
+
+        // Add to the selection handler lookup (for use in onKeyDown)
+        locationSelectedLookup[itemIndex] = locationSelected;
+        // Create and return the option menu item
+        const option = (
+          <Option
+            icon={sessionOptionIcon}
```
Comment (binh-dam-ibigroup) [inline packages/location-field/src/index.tsx:554]: Sort props (multiple instances)

File: packages/location-field/src/options.tsx [hunk 12]
```diff
@@ -21,25 +29,26 @@ export function GeocodedOptionIcon(props) {
...
+  disabled = false,
+  icon = null,
+  isActive = false,
   onClick,
-  title,
-  subTitle,
-  classes
-}) {
+  title = null,
+  subTitle = null,
```
Comment (binh-dam-ibigroup) [inline packages/location-field/src/options.tsx:40]: Sort props.

File: packages/location-field/src/stories/Desktop.story.tsx [hunk 13]
```diff
@@ -10,7 +10,8 @@ import {
   hereGeocoderConfig,
   layerColorMap,
   onLocationSelected,
-  selectedLocation
+  selectedLocation,
+  intlDecorator
```
Comment (binh-dam-ibigroup) [inline packages/location-field/src/stories/Desktop.story.tsx:14]: Sort imports.

File: packages/location-field/src/stories/Mobile.story.tsx [hunk 14]
```diff
@@ -113,33 +112,34 @@ const StyledLocationField = styled(LocationField)`
...
-export const Blank = () => (
+export const Blank = (): JSX.Element => (
   <LocationField
     currentPosition={currentPosition}
     geocoderConfig={geocoderConfig}
     getCurrentPosition={getCurrentPosition}
     locationType="from"
     onLocationSelected={onLocationSelected}
-    static
+    isStatic
```
Comment (binh-dam-ibigroup) [inline packages/location-field/src/stories/Mobile.story.tsx:126]: Sort props after rename (multiple instances)

File: packages/location-field/src/styled.tsx [hunk 15]
```diff
@@ -130,7 +127,11 @@ export const MenuItemA = styled.a`
   white-space: nowrap;
 `;
 
-export const MenuItemHeader = styled.li`
+export const MenuItemHeader = styled.li<{
+  fgColor?: string;
+  bgColor?: string;
+  centeredText?: boolean;
```
Comment (binh-dam-ibigroup) [inline packages/location-field/src/styled.tsx:133]: Sort props

File: packages/location-field/src/styled.tsx [hunk 16]
```diff
@@ -140,76 +141,61 @@ export const MenuItemHeader = styled.li`
...
-  onClick = () => {
-    const { disabled, onClick } = this.props;
+export const MenuItem = ({
+  active = false,
+  centeredText = false,
+  children,
+  // foregroundColor and backgroundColor would be preferred, but React has issues with
+  // these since they are style keywords
+  fgColor = null,
+  bgColor = null,
```
Comment (binh-dam-ibigroup) [inline packages/location-field/src/styled.tsx:159]: Sort props.

File: packages/location-field/src/types.ts [hunk 17]
```diff
@@ -0,0 +1,377 @@
...
+  | "GEOCODE"
+  | "SAVED"
+  | "SESSION"
+  | "STOP";
+export type Location = {
+  id?: string;
+  lat?: number;
+  lon?: number;
+  name?: string;
+  main?: string;
```
Comment (binh-dam-ibigroup) [inline packages/location-field/src/types.ts:17]: Sort attributes.

File: packages/location-field/src/types.ts [hunk 18]
```diff
@@ -0,0 +1,377 @@
...
+  currentPositionIcon?: React.ReactNode;
+  /**
+   * A slot for the icon to display for when the current position is unavailable
+   */
+  currentPositionUnavailableIcon?: React.ReactNode;
+  /**
+   * Allows the component to be rendered with pre-filled results
+   */
+  initialSearchResults?: {
+    type?: string;
```
Comment (binh-dam-ibigroup) [inline packages/location-field/src/types.ts:70]: Sort attributes

File: packages/location-field/src/types.ts [hunk 19]
```diff
@@ -0,0 +1,377 @@
...
+   *
+   * ```js
+   * {
+   *   lat: currentPosition.coords.latitude,
+   *   lon: currentPosition.coords.longitude,
+   *   max: geocoderConfig.maxNearbyStops || 4
+   * }
+   * ```
+   */
+  findNearbyStops?: ({
```
Comment (binh-dam-ibigroup) [inline packages/location-field/src/types.ts:95]: Sort attributes.

File: packages/location-field/src/stories/Desktop.story.tsx [hunk 20]
```diff
@@ -134,7 +136,7 @@ export const WithBadApiKeyHandlesBadAutocomplete = () => (
   />
 );
 
-export const WithCustomResultColorsAndIcons = () => (
+export const WithCustomResultColorsAndIcons = (): JSX.Element => (
```
Comment (binh-dam-ibigroup) [inline packages/location-field/src/stories/Desktop.story.tsx:139]: Maybe rename the story name to also include some custom messages?

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
Looks good, but there is one string concat that escaped the localization filter and that needs to be handled.
Also consider adding custom messages to the story with custom colors and icons. Thanks!

File: packages/location-field/src/index.tsx (hunk 4 above)
Comment (miles-grant-ibigroup) [reply packages/location-field/src/index.tsx]: Addressed in 18d5606109e7dfce7d00d42b42c8b1b6f075a0bc

File: packages/location-field/src/index.tsx (hunk 6 above)
Comment (miles-grant-ibigroup) [reply packages/location-field/src/index.tsx]: Addressed in 18d5606109e7dfce7d00d42b42c8b1b6f075a0bc

File: packages/location-field/src/index.tsx (hunk 8 above)
Comment (miles-grant-ibigroup) [reply packages/location-field/src/index.tsx]: Addressed in 18d5606109e7dfce7d00d42b42c8b1b6f075a0bc

File: packages/location-field/src/index.tsx (hunk 9 above)
Comment (miles-grant-ibigroup) [reply packages/location-field/src/index.tsx]: Addressed in 18d5606109e7dfce7d00d42b42c8b1b6f075a0bc

File: packages/location-field/src/index.tsx (hunk 10 above)
Comment (miles-grant-ibigroup) [reply packages/location-field/src/index.tsx]: Addressed in 18d5606109e7dfce7d00d42b42c8b1b6f075a0bc

File: packages/location-field/src/index.tsx (hunk 11 above)
Comment (miles-grant-ibigroup) [reply packages/location-field/src/index.tsx:550]: Addressed in 18d5606109e7dfce7d00d42b42c8b1b6f075a0bc

File: packages/location-field/src/stories/Desktop.story.tsx (hunk 13 above)
Comment (miles-grant-ibigroup) [reply packages/location-field/src/stories/Desktop.story.tsx]: Addressed in 18d5606109e7dfce7d00d42b42c8b1b6f075a0bc

File: packages/location-field/src/stories/Mobile.story.tsx (hunk 14 above)
Comment (miles-grant-ibigroup) [reply packages/location-field/src/stories/Mobile.story.tsx]: Addressed in 18d5606109e7dfce7d00d42b42c8b1b6f075a0bc

File: packages/location-field/src/styled.tsx (hunk 15 above)
Comment (miles-grant-ibigroup) [reply packages/location-field/src/styled.tsx]: Addressed in 18d5606109e7dfce7d00d42b42c8b1b6f075a0bc

File: packages/location-field/src/styled.tsx (hunk 16 above)
Comment (miles-grant-ibigroup) [reply packages/location-field/src/styled.tsx:158]: ### Addressed in 18d5606109e7dfce7d00d42b42c8b1b6f075a0bc

File: packages/location-field/src/types.ts (hunk 18 above)
Comment (miles-grant-ibigroup) [reply packages/location-field/src/types.ts]: Addressed in 18d5606109e7dfce7d00d42b42c8b1b6f075a0bc

File: packages/location-field/src/index.tsx (hunk 1 above)
Comment (miles-grant-ibigroup) [reply packages/location-field/src/index.tsx]: Addressed in 18d5606109e7dfce7d00d42b42c8b1b6f075a0bc

File: packages/location-field/src/index.tsx (hunk 5 above)
Comment (miles-grant-ibigroup) [reply packages/location-field/src/index.tsx:135]: Nice catch thank you! Fixed in 061fa929c4bd785f52c7dd064ced8a62508b3765

File: packages/location-field/src/index.tsx [hunk 21]
```diff
@@ -0,0 +1,753 @@
...
+  }
+
+  // Add to the selection handler lookup (for use in onKeyDown)
+  locationSelectedLookup[itemIndex] = locationSelected;
+
+  if (!suppressNearby) {
+    // Create and add the option item to the menu items array
+    const currentLocationOption = (
+      <Option
+        icon={optionIcon}
```
Comment (binh-dam-ibigroup) [inline packages/location-field/src/index.tsx:633]: These are the multiple instances I'm referring to!

File: packages/location-field/src/index.tsx [hunk 22]
```diff
@@ -0,0 +1,753 @@
...
+          // If no features found in response, default to empty array.
+          let geocodedFeatures = result?.features;
+          if (!geocodedFeatures) {
+            // Get the Pelias error message if exists.
+            // TODO: determine how other geocoders return error messages.
+            const errorMessage = result?.results?.error?.message;
+            // If the result did not contain a list of features, add special note.
+            message = intl.formatMessage({
+              id: "otpUi.LocationField.geocoderUnreachable"
+            }, { error: errorMessage });
```
Comment (binh-dam-ibigroup) [inline packages/location-field/src/index.tsx:134]: I'd prefer the two-line format as on lines 138-139.

Comment (miles-grant-ibigroup) [reply packages/location-field/src/index.tsx]: Resolved in b4bd8d8c0890c0d8425251b9583773800bc1bc30

**Review by binh-dam-ibigroup — APPROVED** [review]
Thanks for the changes!

**Pull request merged**